netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +824,450 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+    sock_addr_ctx_original = (bpf_sock_addr_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(bpf_sock_addr_t), NET_EBPF_EXTENSION_POOL_TAG);
+    if (sock_addr_ctx_original == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Exit;
+    }
+    memset(sock_addr_ctx, 0, sizeof(bpf_sock_addr_t));
+    memset(sock_addr_ctx_original, 0, sizeof(bpf_sock_addr_t));
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, sock_addr_ctx);
+    *sock_addr_ctx_original = *sock_addr_ctx;
+
+    // Check if this call is intended for us.
+    filter_context = (net_ebpf_extension_sock_addr_wfp_filter_context_t*)filter->context;
+    ASSERT(filter_context != NULL);
+    if (filter_context == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    // If the callout is invoked for v4, then it is safe to simply invoke the eBPF
+    // program from the filter context.
+    // If the callout is invoked for v6:
+    // 1. Check if the destination is v4 mapped v6 address or pure v6 address.
+    // 2. If it is v4 mapped v6 address, then we should procced only if this callout
+    //    is invoked for v4 attach type.
+    // 3. If it is pure v6 address, then we should procced only if this callout is
+    //    invoked for v6 attach type.
+    if (sock_addr_ctx->family == AF_INET6) {
+        if (IN6_IS_ADDR_V4MAPPED((IN6_ADDR*)sock_addr_ctx->user_ip6)) {
+            v4_mapped = true;
+        }
+        if (v4_mapped) {
+            if (!filter_context->v4_attach_type) {
+                // This callout is for v6 attach type, but address is v4 mapped v6 address.
+                // Change action to permit and return.
+                NET_EBPF_EXT_LOG_MESSAGE(
+                    NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                    NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+                    "net_ebpf_extension_sock_addr_redirect_connection_classify: v6 attach type, v4mapped, ignoring");
+                action = FWP_ACTION_PERMIT;
+                goto Exit;
+            }
+        } else if (filter_context->v4_attach_type) {
+            // This callout is for v4 attach type, but address is a pure v6 address.
+            // Change action to permit and return.
+            NET_EBPF_EXT_LOG_MESSAGE(
+                NET_EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+                "net_ebpf_extension_sock_addr_redirect_connection_classify: v4 attach type, purev6, ignoring");
+            action = FWP_ACTION_PERMIT;
+            goto Exit;
+        }
+    }
+
+    attached_client = (net_ebpf_extension_hook_client_t*)filter_context->base.client_context;
+    if (attached_client == NULL) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+    if (!net_ebpf_extension_hook_client_enter_rundown(attached_client)) {
```

Comment:
Shouldnt this be done earlier in the function to avoid unnecessary computations if the client has detached?

Reply:
Before this step, the function is doing 2 things:
1. Allocate and populate sock_addr_ctx. That gives us the AF.
2. Check if this call needs to be processed -- both v4 and v6 connect attach types add filters in `FWPM_LAYER_ALE_CONNECT_REDIRECT_V6`. In the case when there are eBPF programs attached to both the attach types, there will be 2 filters added (though in different sublayers). We need to check here if the invocation is for v4 attach type or v6 attach type.

Since `ExAcquireRundownProtection` performs Interlocked operations, I dont want to call that unless we are sure we should continue processing.